### PR TITLE
Fix #546 - Use sw-precache to make Bramble offline ready

### LINKED ISSUE #546: Use sw-precache for offline support
As part of https://github.com/mozilla/thimble.mozilla.org/issues/1459, we can use https://github.com/GoogleChrome/sw-precache in order to service requests for Bramble from a Service Worker, allowing offline support of the editor (the offline bits of Thimble will have to be done in a second patch).

The way this works is that you specify a set of files to cache via a SW in a grunt task.  I'm trying to come up with that list.  On the surface you might think you can simply specify `dist/*`, but that will result in way too much crap we don't really use (we copy a lot of files over to `dist/` that aren't necessary, but it's easier to copy and just have the right ones served at runtime vs. try and figure out the subset we need at build time).

If you look at what we load on startup, it's the following (which is incomplete, see below):

```
/bramble.js
/index.html
/styles/images/split-view-icons.svg
/styles/images/topcoat-settings-13.svg
/styles/images/live_development_sprites.svg
/styles/images/extension-manager-sprite.svg
/styles/images/no_content_bg.svg
/extensions/default/CSSCodeHints/main.js
/extensions/default/HTMLCodeHints/main.js
/extensions/default/CSSCodeHints/CSSProperties.json
/extensions/default/HTMLCodeHints/HtmlTags.json
/extensions/default/HTMLCodeHints/HtmlAttributes.json
/extensions/default/JavaScriptCodeHints/main.js
/extensions/default/JavaScriptCodeHints/ParameterHintManager.js
/extensions/default/JavaScriptCodeHints/HintUtils.js
/extensions/default/JavaScriptCodeHints/ScopeManager.js
/extensions/default/JavaScriptCodeHints/Session.js
/extensions/default/JavaScriptCodeHints/thirdparty/acorn/acorn.js
/extensions/default/InlineColorEditor/main.js
/extensions/default/JavaScriptCodeHints/HintUtils2.js
/extensions/default/InlineColorEditor/InlineColorEditor.js
/extensions/default/JavaScriptCodeHints/MessageIds.js
/extensions/default/JavaScriptCodeHints/Preferences.js
/extensions/default/JavaScriptQuickEdit/main.js
/extensions/default/JavaScriptCodeHints/thirdparty/acorn/acorn_loose.js
/extensions/default/QuickOpenCSS/main.js
/extensions/default/QuickOpenHTML/main.js
/extensions/default/InlineColorEditor/ColorEditor.js
/extensions/default/QuickOpenJavaScript/main.js
/extensions/default/QuickView/main.js
/extensions/default/WebPlatformDocs/main.js
/extensions/default/CodeFolding/main.js
/extensions/default/bramble/main.js
/extensions/default/Autosave/main.js
/extensions/default/brackets-paste-and-indent/main.js
/extensions/default/WebPlatformDocs/InlineDocsViewer.js
/extensions/default/BrambleUrlCodeHints/main.js
/extensions/default/CodeFolding/Prefs.js
/extensions/default/CodeFolding/foldhelpers/foldgutter.js
/extensions/default/CodeFolding/foldhelpers/foldcode.js
/extensions/default/CodeFolding/foldhelpers/indentFold.js
/extensions/default/bramble/lib/iframe-browser.js
/extensions/default/bramble/lib/UI.js
/extensions/default/bramble/lib/launcher.js
/extensions/default/bramble/nohost/main.js
/extensions/default/bramble/lib/PostMessageTransport.js
/extensions/default/bramble/lib/xhr/XHRHandler.js
/extensions/default/bramble/lib/Theme.js
/extensions/default/bramble/lib/RemoteCommandHandler.js
/extensions/default/bramble/lib/RemoteEvents.js
/extensions/default/InlineColorEditor/thirdparty/tinycolor-min.js
/extensions/default/BrambleUrlCodeHints/camera/index.js
/extensions/default/BrambleUrlCodeHints/selfie.js
/extensions/default/UploadFiles/main.js
/extensions/default/bramble/nohost/HTMLServer.js
/extensions/default/bramble/nohost/StaticServer.js
/extensions/default/bramble/lib/compatibility.js
/extensions/default/bramble/lib/Tutorial.js
/extensions/default/bramble/lib/MouseManager.js
/extensions/default/bramble/lib/LinkManager.js
/extensions/default/CodeFolding/main.css
/extensions/default/BrambleUrlCodeHints/camera/interface.js
/extensions/default/BrambleUrlCodeHints/camera/video.js
/extensions/default/BrambleUrlCodeHints/camera/photo.js
/extensions/default/BrambleUrlCodeHints/camera/utils.js
/extensions/default/BrambleUrlCodeHints/camera-dialog.js
/thirdparty/CodeMirror/addon/fold/brace-fold.js
/thirdparty/CodeMirror/addon/fold/comment-fold.js
/thirdparty/CodeMirror/addon/fold/markdown-fold.js
/extensions/default/UploadFiles/UploadFilesDialog.js
/extensions/default/bramble-move-file/main.js
/extensions/default/bramble-move-file/MoveToDialog.js
/extensions/default/bramble-move-file/MoveUtils.js
/extensions/default/JavaScriptCodeHints/ParameterHintTemplate.html
/extensions/default/JavaScriptCodeHints/keyboard.json
/extensions/default/JavaScriptCodeHints/thirdparty/tern/defs/ecma5.json
/extensions/default/JavaScriptCodeHints/thirdparty/tern/defs/browser.json
/extensions/default/QuickView/QuickViewTemplate.html
/extensions/default/WebPlatformDocs/InlineDocsViewer.html
/extensions/default/InlineColorEditor/ColorEditorTemplate.html
/extensions/default/BrambleUrlCodeHints/data.json
/extensions/default/BrambleUrlCodeHints/camera/selfieWidget.html
/extensions/default/BrambleUrlCodeHints/dialog.html
/extensions/default/bramble/lib/PostMessageTransportRemote.js
/extensions/default/bramble/lib/xhr/XHRShim.js
/extensions/default/bramble/lib/Mobile.html
/extensions/default/bramble/lib/MouseManagerRemote.js
/extensions/default/bramble/lib/LinkManagerRemote.js
/extensions/default/UploadFiles/htmlContent/upload-files-dialog.html
/extensions/default/bramble-move-file/htmlContent/move-to-dialog.html
/extensions/default/bramble-move-file/htmlContent/directory-tree.html
/extensions/default/InlineColorEditor/css/main.css
/extensions/default/QuickView/QuickView.css
/extensions/default/WebPlatformDocs/WebPlatformDocs.css
/extensions/default/UploadFiles/styles.css
/extensions/default/bramble/stylesheets/style.css
/extensions/default/bramble/stylesheets/sidebarTheme.css
/extensions/default/bramble-move-file/styles/style.css
/extensions/default/BrambleUrlCodeHints/style.css
/extensions/default/CSSCodeHints/styles/brackets-css-hints.css
/extensions/default/JavaScriptCodeHints/styles/brackets-js-hints.css
/extensions/default/bramble/stylesheets/lightTheme.css
/extensions/default/JavaScriptCodeHints/tern-worker.js
/extensions/default/JavaScriptCodeHints/thirdparty/tern/lib/tern.js
/extensions/default/JavaScriptCodeHints/thirdparty/tern/lib/infer.js
/extensions/default/JavaScriptCodeHints/thirdparty/acorn/util/walk.js
/extensions/default/JavaScriptCodeHints/thirdparty/tern/lib/def.js
/extensions/default/JavaScriptCodeHints/thirdparty/tern/lib/signal.js
```

There are still some files that we'll load at runtime on first use (e.g., images, stylesheets, json files, ...) that we have to specify in this list.  I need to sort out how to deal with this.

### PR DESCRIPTION
**Pull request by humphd**
With this patch, I can run Bramble offline.  It's really cool.  The major downside to this is that it requires us to keep a static list of files in order to "prime" the cache, and register handlers for the Service Worker to respond to when URLs are requested.  I don't see us adding a lot of files, but it's something we'll have to keep an eye on.  I need to document this, too. 

This probably needs another bit of tidying up, but it's pretty close.

cc @gideonthomas

### CODE REVIEW

File: Gruntfile.js [hunk 1]
```diff
@@ -575,6 +584,29 @@ module.exports = function (grunt) {
...
+
+        var config = {
+            cacheId: 'bramble-cache::' + Date.now(),
+            logger: grunt.log.writeln,
+            staticFileGlobs: files,
+            stripPrefix: 'dist/',
+            ignoreUrlParametersMatching: [/./]
+        };
+
+        swPrecache.write(Path.join(rootDir, 'bramble-sw.js'), config, function(err) {
```
Comment (gideonthomas) [inline Gruntfile.js]: shouldn't this write to `dist/`?

File: Gruntfile.js [hunk 2]
```diff
@@ -612,7 +644,8 @@ module.exports = function (grunt) {
         'replace:ternDefs',
         'requirejs:iframe',
         'exec:localize-dist',
-        'uglify'
+        'uglify',
+        'swPrecache'
```
Comment (gideonthomas) [inline Gruntfile.js]: Wouldn't you want this to run after the `compress` task since production runs `grunt build-browser-compressed`?

File: package.json [hunk 3]
```diff
@@ -43,7 +43,6 @@
         "load-grunt-tasks": "0.2.0",
         "q": "0.9.2",
         "semver": "^4.1.0",
-        "jshint": "2.1.4",
```
Comment (gideonthomas) [inline package.json]: Why this change?

File: Gruntfile.js [hunk 4]
```diff
@@ -575,6 +584,29 @@ module.exports = function (grunt) {
         ]);
     });
 
+    grunt.registerMultiTask('swPrecache', function() {
+        var done = this.async();
+        var rootDir = this.data.rootDir;
+        var files =  (function() {
```
Comment (gideonthomas) [inline Gruntfile.js]: Does this rely on globbing patterns? Can't we just do something like `[ "dist/**/*" ]`?

**Comment by gideonthomas** [comment]
Few questions otherwise looks good to merge

File: Gruntfile.js (hunk 4 above)
Comment (humphd) [inline Gruntfile.js]: Yes, you could do this, if we were smarter about how we copied things to `dist/`.  But we aren't smart.  Everything that you tell the SW to precache, it will download and cache.  So if you ask for more than you need, you take a big hit on network installation on first run.  If you look at what's in `dist/`, we copy all kinds of crap for CodeMirror and other things.  We could fix that, but it's going to basically involve doing what I'm doing here: keeping a static list of things we care about.  So I'm not sure it's worth fixing.  There's no harm copying them to `dist/`; but we don't really want things getting downloaded/cached on the browser that we know we'll never use.

File: Gruntfile.js (hunk 1 above)
Comment (humphd) [inline Gruntfile.js]: That's what it's doing, see https://github.com/mozilla/brackets/pull/547/files/46deab8a7cddc92bbcb6683675ec4e2c5aa65b1b#diff-35b4a816e0441e6a375cd925af50752cR528.

File: Gruntfile.js (hunk 2 above)
Comment (humphd) [inline Gruntfile.js]: Yeah, you're right.  I'll fix.

**Comment by humphd** [comment]
@gideonthomas, I've updated things, you OK with it now r?

**Comment by gideonthomas** [comment]
r+
